docs(cli/verify): initial version (#2744)
latticexyz/mud: docs/pages/cli/verify.mdx

The best time to perform verification is immediately after [deployment](./deploy).
Verification only works with the original source code, compiler options, and compiled artifacts.
Otherwise, the generated bytecode is different and therefore verification fails.

## Using the command

1. If using Blockscout, you need to specify the API key:

   ```sh copy
   export ETHERSCAN_API_KEY=<value goes here>
   ```

1. Run the command.

   ```sh copy
   pnpm mud verify --worldAddress <address here> --verifier <verifier> --rpc <RPC to the blockchain>
   ```

   These are the command line options you can specify on `mud verify`:

   | Option           | Meaning                                                | Type    | Default value                |
   | ---------------- | ------------------------------------------------------ | ------- | ---------------------------- |
   | `--worldAddress` | Verify the contracts of the World at the given address | string  | none, error if unspecified   |
   | `--configPath`   | Path to the config file                                | string  | `mud.config.ts`              |

>    | `--verifier`     | The verifier to use (`sourcify`, or `blockscout`)      | string  |                              |

Reviewer: The default verifier will be `sourcify` in the new PR

```suggestion
   | `--verifier`     | The verifier to use (`sourcify`, or `blockscout`)      | string  | sourcify |
```

   | `--verifierUrl`  | URL to use to access the verifier's API                | string  | depends on the verifier used |
   | `--profile`      | The foundry profile to use                             | string  | `local`                      |
   | `--srcDir`       | Source directory                                       | string  | Foundry `src` directory      |
   | `--version`      | Show version number                                    | boolean | `false`                      |

   Note that it takes block explorers some time to process new contracts.
   You should wait 5-10 minutes past deployment before you verify the contract on Etherscan and/or Blockscout.

## Examples

- To verify the world at that address on Holesky on [sourcify](https://sourcify.dev/):

  ```sh
  WORLD_ADDRESS=0x816038e244ff78b86a5e7dec9bf281634fb6d2a2
  pnpm mud verify --rpc https://ethereum-holesky.publicnode.com --worldAddress $WORLD_ADDRESS --verifier sourcify
  ```

> - To verify on Holesky on [Blockscout](https://www.blockscout.com/):

Reviewer: Worth noting here (or somewhere else) that as of https://github.com/latticexyz/mud/pull/2750 we can currently only verify the "local" contracts with `blockscout` (ie. user-defined systems, not the store, world, world-modules contracts)

# mud verify

This command uploads the source code for the deployed contracts to a public repository, such as a block explorer.

> The repository then compiles the source code and verifies that the compiled version is identical to what is deployed on-chain.

Reviewer 2: ```suggestion
The repository then compiles the source code and verifies that the compiled version is identical to what is deployed onchain.
```

## Using the command

> 1. Before you run `mud verify` you need to create a [Foundry profile](https://book.getfoundry.sh/reference/config/overview).

Reviewer 2: you don't _need_ to, it just makes it a bit easier to pass in `--profile` to set various foundry options rather than individual things like `--rpc`

   Change to the `packages/contract` directory and edit `foundry.toml` to add a profile with the `eth_rpc_url` for your network (if needed).
   For example, add these lines to add Holesky.

   ```toml copy
   [profile.holesky]
   eth_rpc_url = "https://ethereum-holesky.publicnode.com"
   ```

   Note that this blockchain has to be supported by the verifier.

1. If using Blockscout, you need to specify the API key:

   ```sh copy
   export ETHERSCAN_API_KEY=<value goes here>

>    ```

Reviewer 3: i actually think blockscout doesn't require an api key (but it might be that forge expects _something_ to be provided as an an api key)

Reviewer 2: no blockscout API key afaik, only need to provide an etherscan API key if you use `--verifier etherscan` or no `--verifier` (forge defaults to etherscan)

Author: 1. We (@yonadaaa and me) decided not to document Etherscan verification at this time because neither of us managed to get it working.

2. If I understand https://docs.blockscout.com/for-users/api#api-keys correctly, it means that Blockscout will serve up to 50 calls without a key every second. Yours may or may not be one of them. But with an API key you get 10 calls a second for yourself.

> 1. Run the command.

Reviewer 2: I think it's worth noting here that verify will only work with the original source + compiler options + compiled artifacts all match the onchain bytecode. So verifying should often happen immediately after deploying and before any further edits, package updates, etc. are done.

   ```sh copy

>    pnpm mud verify --worldAddress <address here> --verifier <verifier> --profle <profile>

Reviewer 2: should we use `--profile holesky` here since we used holesky provide in the example above?

Author: No, here we explain what to do so we need to be precise. The examples are below

   ```

   These are the command line options you can specify on `mud verify`:

   | Option           | Meaning                                                  | Type    | Default value                |
   | ---------------- | -------------------------------------------------------- | ------- | ---------------------------- |

>    | `--worldAddress` | Verify the contracts of the `World` at the given address | string  | none, error if unspecified   |

Reviewer 2: ```suggestion
   | `--worldAddress` | Verify the contracts of the World at the given address | string  | none, error if unspecified   |
```

>    | `--version`      | Show version number                                      | boolean | `false`                      |

Reviewer 2: no need to document this, this isn't specific to this command but for the whole CLI

Author: I think we should document it, for completeness.

Reviewer 2: disagree, it's just noise here

   | `--configPath`   | Path to the config file                                  | string  | `mud.config.ts`              |
   | `--verifier`     | The verifier to use (`sourcify`, or `blockscout`)        | string  |                              |

>    | `--verifierURL`  | URL to use to access the verifier's API                  | string  | depends on the verifier used |

Reviewer 2: ```suggestion
   | `--verifierUrl`  | URL to use to access the verifier's API                  | string  | depends on the verifier used |
```

   | `--rpc`          | URL to the blockchain                                  | string  |                              |
   | `--srcDir`       | Source directory                                       | string  | Foundry `src` directory      |
   | `--version`      | Show version number                                    | boolean | `false`                      |

   Note that it takes block explorers some time to process new contracts.
   You should wait 5-10 minutes past deployment before you verify the contract on Etherscan and/or Blockscout.

## Examples

- To verify the world at that address on Holesky on [sourcify](https://sourcify.dev/):

  ```sh
  WORLD_ADDRESS=0x816038e244ff78b86a5e7dec9bf281634fb6d2a2
  HOLESKY_RPC=https://ethereum-holesky.publicnode.com
  pnpm mud verify --rpc $HOLESKY_RPC --worldAddress $WORLD_ADDRESS
  ```

- To verify a contract on Holesky for users of [Blockscout](https://www.blockscout.com/):

  ```sh
  WORLD_ADDRESS=0x816038e244ff78b86a5e7dec9bf281634fb6d2a2
  HOLESKY_RPC=https://ethereum-holesky.publicnode.com
  export ETHERSCAN_API_KEY= ** blockscout API key goes here **

>   pnpm mud verify --rpc HOLESKY_RPC --worldAddress $WORLD_ADDRESS --verifier blockscout --verifier-url 'https://eth-holesky.blockscout.com/api?'

Reviewer 2: ```suggestion
  pnpm mud verify --rpc $HOLESKY_RPC --worldAddress $WORLD_ADDRESS --verifier blockscout --verifier-url 'https://eth-holesky.blockscout.com/api?'
```
is the `?` at the end of the url necessary?

Author: I verified it isn't. I'll remove it.
